### ISSUE TITLE
Randomly generated quests

### ISSUE THREAD

**Issue by jeremyckahn** [jeremyckahn/farmhand#151]
Random quest ideas:

  - Grow X crops in Y timeframe

Earn quest coins to unlock buffs like auto watering, auto fertilizing, cow feed purchasing, etc.

--------------------------

This idea needs more refinement!

**Comment by karlerik-simon** [comment]
Sounds like a fun issue! Might I have a go at it? :)

**Comment by jeremyckahn** [comment]
@karlerik-simon sure! Thanks for offering to help.

Before development starts on this, we'll need to flesh out the concept a bit more. The goal of this mechanic is to keep the game interesting indefinitely (or at least as long as possible). I haven't had a chance to really figure out how quests might work in practice, aside from them being procedurally generated. I think we need to start with defining types of quests, how they get defined procedurally, and what the reward structure is. Do you have specific ideas on how quests might work in Farmhand?

Just to set expectations, I think all we need to start with is the groundwork for a quest system. Like anything in this project, we should work on this iteratively and deliver meaningful milestones as soon as they're stable and cohesive, not wait for some grand vision to be complete before shipping.

This is a larger design undertaking that I am very happy to have help with!

**Comment by karlerik-simon** [comment]
Totally agree! 

In regards to what types of quests there should be I was thinking about things like **grow 40 carrots**, **reach level 5**, **increase the field by xx** and so on. And that the further the player is in the game, the harder the quests. 

In regars to rewards, perhaps either coins or items. ex. a scarecrow and $200 for a quest, and some quests only rewarding coins or items

What do you think?

As you mentioned @jeremyckahn, this is a big undertaking, and I was thinking of drafting something in photoshop of how I'd imagine it. Also, we need to make a design that blends in into the UI, or maybe have a questionmark where the quest(s) are listed?

**Comment by karlerik-simon** [comment]
Maybe also have timed quests, like sell xx of xxxx in 10 minutes?

**Comment by jeremyckahn** [comment]
> In regards to what types of quests there should be I was thinking about things like **grow 40 carrots**, **reach level 5**, **increase the field by xx** and so on.

"Static" goals like "reach level 5" and "increase the field by xx" seem to be a better fit for achievements, because they can be only be done once.

> And that the further the player is in the game, the harder the quests.

I really like the idea of scaling the challenge as the user progresses! Maybe we could scale it to their level in the game? We may also want to consider deferring quest scaling to a later iteration for simplicity's sake.

> In regards to rewards, perhaps either coins or items. ex. a scarecrow and $200 for a quest, and some quests only rewarding coins or items

I like this idea! I initially hinted at a "coin" concept, but I realize now that it may be over-complicating the first iteration. Let's stick with money and item rewards and explore "coins" later.

> I was thinking of drafting something in photoshop of how I'd imagine it.

Great idea! 👍 

> Also, we need to make a design that blends in into the UI, or maybe have a questionmark where the quest(s) are listed?

I think it makes sense to restrict quest UI to a modal screen similar to the Price Events or Farm Stats screens. We can inform the user about quest updates with the notification/toast system. And similar to those, we could add a new modal trigger button to the Price Events/Farm Stats/etc. button array.

> Maybe also have timed quests, like sell xx of xxxx in 10 minutes?

I'd like to avoid real time-based quests, as Farmhand is a fundamentally turn-based game (with the exception of multiplayer mechanics). However, I think giving day limits to quests would fit very well (e.g. "harvest 50 carrots in 4 days").

There are some really great ideas here. It's exciting to see this finally take shape! 😁

**Comment by jeremyckahn** [comment]
[On Discord](https://discord.com/channels/714539345050075176/714539345637408793/1060635720278085753), M8theone suggested giving buffs (like increased sale values) as rewards. I think this is an excellent idea, but we may want to save this for a later iteration to keep the initial scope small.

**Comment by karlerik-simon** [comment]
Very great points! I was also thinking about a modal for UI, to not clutter the main UI to much. 

> M8theone suggested giving buffs (like increased sale values) as rewards

I really like this! Could be great if implemented correctly. I agree on starting with coins and items as rewards, but keeping the issue open for further implementation of buffs and such. Regarding permanent buffs I think we're on the same page, but as you said, worth exploring later on!

**Comment by karlerik-simon** [comment]
Thanks for assigning me! I am close to finishing autosaving, after which I will start with a design draft for this! Great fun to work on this game!

**Comment by jeremyckahn** [comment]
Awesome, I'm glad you're enjoying it! My goal is to keep Farmhand as much fun to develop as it is to play. :)

**Comment by jeremyckahn** [comment]
I raised [a concern in Discord](https://discord.com/channels/714539345050075176/714539345637408793/1060777228926996502) about the "quest" terminology for this feature. I think the concept fits well, but "quest" semantics seem like an awkward fit for Farmhand's overall aesthetic and theme. M8theone had this to say:

> Maybe something like requests at a community board, kind of like Hayday’s system where you can donate crops and stuff to get money and energy

I really like the idea of having players do random tasks requested by a fictional community. This gives us room to grow the concept within the Farmhand universe and enable more interesting gameplay mechanics down the road.

Previously we had discussed putting the quest UI in a modal. However, this seems like a good opportunity to think bigger: How about a new page (similar to the Field, Cow Pen, or Workshop) for "Community?" To start with, that page would just contain the "quest" UI (though I suggest calling it something like "Help Wanted" to convey that the player is doing work for a fictional community). Over time, we could expand the Community page with tabs (similarly to the Shop and Workshop pages) with new gameplay mechanics.

This is a bit of an increase in scope, but I think it will set us up for more interesting possibilities down the road. What do you think, @karlerik-simon?

**Comment by jeremyckahn** [comment]
> One kind of "out there" idea with this is that maybe in the future we could add a "community score" where the more you help out the more "street creed" you earn which in turn makes certain crops easier to buy. Just a thought though, not at all relevant at the moment.

I like it! Something also to keep in mind: We should definitely track metrics for quests/jobs and display them in the Farm Stats page. We could use them for achievement goals, too!

**Comment by karlerik-simon** [comment]
> Something also to keep in mind: We should definitely track metrics for quests/jobs and display them in the Farm Stats page. We could use them for achievement goals, too!

Certainly! I think that would be great. I really like the new "quest" concept. A thought I had though about the market concept. Right now you can sell you foraged items for money, what difference should there be to the community market in that you are still selling items for a reward?

**Comment by jeremyckahn** [comment]
> Right now you can sell your foraged items for money, what difference should there be to the community market in that you are still selling items for a reward?

By "foraged items," I'm assuming you're referring to mined ores and field weeds. Is that right?

I'm not sure that the sale of foraged items would necessarily have an effect on the community market. We could design a mechanic around that, but it doesn't seem necessary for a V1 of this feature to me. Is there something about foraged items that you think should be taken into account with the design of the community market?

**Comment by karlerik-simon** [comment]
Exactly! No not really. Maybe in the future with the community score concept, not as you said. Not in a V1. I'll start drafting something up!

**Comment by lstebner** [comment]
it's awesome to see this taking shape! this is a great discussion so far. i especially like the "community board"/help wanted approach and the fictional community. 

when i think of something like quests i imagine that instead of the goals being related to selling that the player actually chooses to deliberately fulfill the request instead, and the reward for the quest would be a monetary value that is higher than simply selling them. though thinking about quests has also made me want to migrate farmhand to an experience based leveling system so that quests can reward the player both experience and money. that's a separate issue/task though that we can think about.

I imagine quests could be something like "Bob needs 30 carrots and 20 onions to make his famous carrot stew. He'll pay $100 for the items." (don't take that value literally, i didn't do any math hah), or "Susan is building a new fence and needs 500 stones", and then instead of selling those items the player actually presses a button in this new quest UI when they have those items in their inventory like "fulfill request" and the items are removed in return for the reward. and maybe the reward isn't always money, but sometimes other goods like a trade.

I definitely see some obvious achievements that could come from this too for completing quests. we could also theme them around the community concept like "friend of the people" which could be rewarded for completing some number of quests aka "helping <number> people around town."

Another idea on my mind is to have quests unlock after the player has played for a bit similarly to mining. maybe around level 5 they become available to sorta expand the things the player can do after they're used to the initial game loop. 

One last thing, I think we should make sure that quest goals are not just focused around growing crops but also mining, crafting, and even weeding. They could be a good way to encourage the player to do a variety of things, and give extra purpose to things like stone and weeds that are low in value to sell. It will be fun to come up with some of the "stories" for these!

These are just ideas of course so feel free to pick them apart. I think this feature is going to add a lot of fun stuff to do in the game!

**Comment by jeremyckahn** [comment]
@lstebner these are some great ideas! What you're describing is generally where I'd like the "Community Board" system (that's what I'm going to call the quest system for now) to go long term. And it's important to have this more fleshed-out vision articulated here just like you have it so we know what we're building towards!

Like anything with Farmhand, I think we'll need to start with the simplest possible expression of an idea so we can ship a V1 as quickly as possible. That'll enable us to rapidly test the concept with real players and validate our assumptions. From there we can layer on enhancements to the core idea while having something to show for it with each iteration. I think @karlerik-simon's [concept](https://github.com/jeremyckahn/farmhand/issues/151#issuecomment-1372385473) is closer in scope to what a Community Board V1 should look like. Adding narrative elements (like characters and motivating goals) as @lstebner described would fit nicely as a bolt-on enhancement to that core.

> One last thing, I think we should make sure that quest goals are not just focused around growing crops but also mining, crafting, and even weeding. They could be a good way to encourage the player to do a variety of things, and give extra purpose to things like stone and weeds that are low in value to sell. It will be fun to come up with some of the "stories" for these!

Definitely! This will be a great way of getting more mileage out of some of the less-integral items in the game.
